goty-server/src/test/kotlin/com/aleinin/goty/properties/PropertiesControllerTest.kt

```diff
@@ -0,0 +1,124 @@
+    lateinit var propertiesRepository: PropertiesRepository
+
+    @Autowired
+    lateinit var defaultProperties: DefaultProperties
+
+    @Autowired
+    lateinit var objectMapper: ObjectMapper
+
+    @Captor
+    lateinit var documentCaptor: ArgumentCaptor<PropertiesDocument>
+
+    private val oneDayInSeconds = (24 * 60 * 60).toLong()
+
+    private val mockPropertiesDocument = PropertiesDocument(
+        id = "id",
+        tiePoints = listOf(3, 2, 1),
+        deadlineDate = Date.from(Instant.now().plusSeconds(oneDayInSeconds)),
+        zoneId = ZoneId.systemDefault(),
+        hasGiveaway = true,
+        giveawayAmountUSD = 0
+    )
+
+    @BeforeEach
+    fun setup() {
+        val propertiesService = PropertiesService(propertiesRepository, defaultProperties)
+        mockMvc = MockMvcBuilders.standaloneSetup(PropertiesController(propertiesService)).build()
+    }
+
+    @Test
+    fun `Should get stored properties if available`() {
+        whenever(propertiesDocumentRepository.findById(any())).thenReturn(Optional.of(mockPropertiesDocument))
+        val expected = Properties(
+            tiePoints = mockPropertiesDocument.tiePoints,
+            deadline = mockPropertiesDocument.deadlineDate.toInstant().atZone(mockPropertiesDocument.zoneId),
+            hasGiveaway = mockPropertiesDocument.hasGiveaway,
+            giveawayAmountUSD = mockPropertiesDocument.giveawayAmountUSD
+        )
+        val expectedJsonContent = objectMapper.writeValueAsString(expected)
+        mockMvc.perform(get("/properties"))
+            .andExpect(status().isOk)
+            .andExpect(content().json(expectedJsonContent, true))
+    }
+
+    @Test
+    fun `Should get the default properties if none are stored`() {
+        whenever(propertiesDocumentRepository.findById(any())).thenReturn(Optional.empty())
+        val expected = defaultProperties.toProperties()
+        val expectedJsonContent = objectMapper.writeValueAsString(expected)
+        mockMvc.perform(get("/properties"))
+            .andExpect(status().isOk)
+            .andExpect(content().json(expectedJsonContent, true))
+    }
+
+    @Test
+    fun `Should replace the properties`() {
+        val request = Properties(
+            tiePoints = listOf(15, 13, 11),
+            deadline = tomorrow().truncatedTo(ChronoUnit.SECONDS),
+            hasGiveaway = false,
+            giveawayAmountUSD = 15
+        )
+        val expectedDocument = PropertiesDocument(
+            id = PropertiesRepository.PROPERTIES_ID,
+            tiePoints = request.tiePoints,
+            deadlineDate = Date.from(request.deadline.toInstant()),
+            zoneId = request.deadline.offset,
+            hasGiveaway = request.hasGiveaway,
+            giveawayAmountUSD = request.giveawayAmountUSD
+        )
+        val requestAsJsonString = objectMapper.writeValueAsString(request)
+        whenever(propertiesDocumentRepository.save(Mockito.any(PropertiesDocument::class.java))).thenReturn(expectedDocument)
```

Comment:
Couldn't figure this one out. mockito-kotlin's any blew up on the JPA saves. Their any also doesn't support class args like Mockito's any

Reply:
My only guess is since `PropertiesDocumentRepository` is a regular Mockito mock there must be some kind of incompatibility. I did get this to work by doing the following:

```kotlin
whenever<Any?>(propertiesDocumentRepository.save(any())).thenReturn(expectedDocument)
```

TBH, I think what you have is just fine.